Caching

Ebstein opened the pull request:
So many changes happened. I have resolved the following:

1) Parsing of the data and caching of it is now different. We remember all the categories to which every event belongs in the format ,N,NN,NN, where N are digits. Hence, we have a function that creates such a string within EventsParseForDateWithinCategory and we have a static function that retrieves first non null category to which such an event belongs to pick the right color of the rectangle/blob available in the same class. 

2) The events now works when no data is cached at all and does not crash if internet is not available. That required creating a callback loop between the inherited adapter (EventsAdapterForLists) and successing classes (EventsListAdapter and EventsCalendarEventList) so that when the asynctask finishes loading data and parsing it, they are displayed immediately. It is quite elegant and works fine. The only glitch with this approach is that the dialog window in listView will disappear for a moment between downloading the data for current and next month. 

The overridden "notifyDataSetChanged" in each of the successor classes will update the datasets and display them. 

3) I have commented many parts of the code more extensively. Especially the adapter classes, so that there is no confusion about the callback loop that now exists. I hope it is clear.

4) I have extracted one of the functions operating on months to DateFormater from EventsParseForDateWithinCategory. 


Oh yeah, and I forgot to sign into a shift. will add the time manually.

Ebstein commented:
We need a good schema for updating the cache. As I noticed over the snow-emergency days, the calendar is actually updated (especially when it comes to delayed and postponed events!). When do we update the data set? Can we have a flag from API that indicates which event has been updated? Or perhaps can we query the API for events updated after certain date and update only those? That would be super helpful!  @jasonkliu @jeminlee 

For now I am caching events from current month and for all the months up to +6 in the future. It takes a while. 

@jeminlee @jasonkliu : Should I really disable the "month back" button in the calendar if we are viewing current month? It seems like a waste and honestly, if someone wants to see a past event, why not allow them to do so?

jasonkliu commented:
One solution to this is to provide an update flag similar to the one in #46:
We can have the app query the json and find something like a date which we can check if the cache is updated to - if not, it tells the app to manually update the cache.
It provides an additional level of abstraction so that we can push data to the app without having to republish it.

jasonkliu commented on src/edu/yalestc/yalepublic/Cache/CalendarCache.java:
```diff
@@ -81,11 +81,11 @@ private void updatePreferences(int currentYear, int currentMonth) {
         int mDay = 1;
         createEventPreferences.putInt("dateCached", Integer.parseInt(DateFormater.calendarDateToEventsParseForDate(mYear, mMonth, mDay)));
-        int topMonth = mMonth + 2;
+        int topMonth = mMonth + 6;
```
You should write some sort of comment detailing what these are: `topMonth` and `botMonth` seem a bit important :banana:

Ebstein commented:
@jasonkliu @jeminlee Oh, just a disclaimer at the beginning: I could not log in into shifts for the most part of my working time today since Yale Secure on Linux is just not working well today. I will log automatically. I have almost finished all that I was supposed to do in regards to events. The following have changed: 

1) Introduced final static variables in caching so that we can easily change the number of months cached

2) introduced comments in all classes. This should help all of us soo much. I will go through all the classes and do quick reformatting / commenting to make it all look good.

4) The parser now can return a list of days on which there are events within given category. Complexity: n. It's not very inefficient in the sense that we already once go through the events (when parsing after download), but I did not see a decrease in speed. n is typically 200-300. 

5) Added method to retrieve the days with events within given category from db. Again, this is not optimal for time, but for space. We do not introduce additional tables to take care of this, we just go through all the events (n ~= 200-300) and collect the days on which we have events.

6) The calendar now shows blobs next to dates if there are events present on a given day. This required linking the grid adapter with list adapter and introducing a flag in list adapter, so that we know when to update the grid adapter. It is not the most elegant way to do it, but it is as good as it gets with the current structure of this part of the app. Had I known from the beginning that we wanted this feature, the classes would probably be structured differently. I would be happy to rewrite part of them after the deadline.

7) The app does not crash when the tile selected in former month is out of bounds in current month. When going to new month we are selecting either day 1 of given month, or current day if it is within bounds (if current month is selected)

8) Repaired caching of the next 6 months - now we actually are doing it. before there was a glitch

9) It is still possible to peek into former months, but they are not cached. I don't think limiting the calendar is a good idea.

10) I could not push the caching part outside of mainActivity, because currently there is no splash screen (anyone working on it?)

I hope I did not forget about anything. In general, the classes for calendar view got pretty complex in that they all interact with each other in ways that normally are not too apparent. As said before, I'll be happy to rewrite them after the demo, but for now it all works. Until I forget how it works (I could figure it out from comments but overall it would take a bit :-))

jasonkliu commented:
I'm running into an issue where the "all" view returns all of the events correctly, but each separate category has no events in it.  The List View of each category is also blank.
(I just compiled and built the application then installed it from Android Studio)

Ebstein commented:
That's weird, it works for me...

jasonkliu commented:
Try fully uninstalling and reinstalling the application? That should show the same result.

Ebstein commented:
It works.

Ebstein commented:
repairing the "static" inner fragment. Here is a nice explaination of why we should do it: 

http://stackoverflow.com/questions/15571010/fragment-inner-class-should-be-static

Which also basically says that we are fine... but to be better off in the future...

jasonkliu commented:
It appears that I had an old version of the application which didn't correctly query for the categories.  Fully uninstalling and reinstalling fixed the issue.

Which brings up the question, should we allow the user to manually re-cache data? Maybe a thought for the future.

Ebstein commented:
I'm against that. But we should have some way of forcing an update remotely. Maybe similar system as with the changing icon?
